keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,208 @@
+- Not changing the behavior of other label selectors, such as `Replicases`, `Daemonsets`, etc.
+
+ ## Proposal
+
+ ### User Stories
+
+ As an application developer, I want my application pods to be scheduled onto
+one node that has pod with the "foo" tag and the tag value between "20" and "30".
+
+- if we use the original `Exists` operator to implement, then we will write affinity 
+like this:
+ ```go
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Exists
+            values:
+            - 20
+       - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Exists
+            values:
+            - 21
+       - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Exists
+            values:
+            - 22
+          ...
+       - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Exists
+            values:
+            - 29
+       - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Exists
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+- if we use the original `In` operator to implement, then we will write affinity 
+like this:
+ ```go
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ Both are not ideal solutions. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```go
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. However, the performance of the entire scheduler has been greatly improved(https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
```

Comment:
Would be good to mention performance/benchmark tests will be covered in roll out plans.

And also backward compatibility test is mandatory.

Reply:
Done